src/test/metrics_test.cpp

```diff
@@ -0,0 +1,808 @@
+    registry.remove(family);
+    GTEST_SKIP_("Skipped: Using metric after removal is undefined behavior");
+    metric->observe(0.01);
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name_sum{label=\"value\"}")));
+}
+
+TEST(MetricsHistogram, RemoveMetric) {
+    MetricRegistry registry;
+    auto family = registry.createFamily<MetricHistogram>("name", "desc");
+    auto metric1 = family->addMetric({{"label", "value"}}, {10});
+    auto metric2 = family->addMetric({{"other", "data"}}, {10});
+    EXPECT_THAT(registry.collect(), HasSubstr("# HELP name"));
+    EXPECT_THAT(registry.collect(), HasSubstr("# TYPE name"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"10\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"+Inf\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_sum{label=\"value\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_count{label=\"value\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{other=\"data\",le=\"10\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{other=\"data\",le=\"+Inf\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_sum{other=\"data\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_count{other=\"data\"} 0\n"));
+    family->remove(metric1);
+    EXPECT_THAT(registry.collect(), HasSubstr("# HELP name"));
+    EXPECT_THAT(registry.collect(), HasSubstr("# TYPE name"));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name_bucket{label=\"value\",le=\"10\"}")));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name_bucket{label=\"value\",le=\"+Inf\"}")));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name_sum{label=\"value\"}")));
+    EXPECT_THAT(registry.collect(), Not(HasSubstr("name_count{label=\"value\"}")));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{other=\"data\",le=\"10\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{other=\"data\",le=\"+Inf\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_sum{other=\"data\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_count{other=\"data\"} 0\n"));
+}
+
+TEST(MetricsHistogram, RemoveLastMetricRemovesFamily) {
+    MetricRegistry registry;
+    auto family = registry.createFamily<MetricHistogram>("name", "desc");
+    auto metric = family->addMetric({{"label", "value"}}, {10});
+    EXPECT_THAT(registry.collect(), HasSubstr("# HELP name"));
+    EXPECT_THAT(registry.collect(), HasSubstr("# TYPE name"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"10\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_bucket{label=\"value\",le=\"+Inf\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_sum{label=\"value\"} 0\n"));
+    EXPECT_THAT(registry.collect(), HasSubstr("name_count{label=\"value\"} 0\n"));
```

Comment:
Do we really want to test that metric appears after creation here? We test it in other tests

Reply:
Removed